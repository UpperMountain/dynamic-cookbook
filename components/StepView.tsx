import React from "react";
import { View, Text, TouchableOpacity } from "react-native";
import { Step } from "../lib/graph";
import LeftLine from "./LeftLine";
import Markdown from "react-native-markdown-renderer";
import Padded from "./Padded";
import Card from "./Card";
import { padding } from "../lib/theme";
import { MaterialIcons } from "@expo/vector-icons";

const topSize = 40;

interface CircledProps {
  children: string;
}
function Circled(props: CircledProps) {
  const { children } = props;
  return (
    <View
      style={{
        width: topSize,
        height: topSize,
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
        borderWidth: 1,
        borderColor: "black",
        borderRadius: 25
      }}
    >
      <Text style={{ fontSize: 18 }}>{children}</Text>
    </View>
  );
}

function stringFormatInterval(seconds: number) {
  let minutes: number | null = null;
  let hours: number | null = null;
  if (seconds > 60) {
    minutes = Math.floor(seconds / 60);
    seconds = seconds % 60;
  }
  if (minutes && minutes > 60) {
    hours = Math.floor(minutes / 60);
    minutes = minutes % 60;
  }

  let out = "";
  if (hours) out += hours.toString().padStart(2, "0") + ":";
  if (minutes) out += minutes.toString().padStart(2, "0") + ":";
  out += seconds.toString().padStart(2, "0");

  if (!hours && !minutes) {
    out += " s";
  }

  return out;
}

export interface StepActionProps {
  // Time remaining, or undefined if not started
  remaining?: number;

  // if this card is, or leads to, a timer
  timer?: boolean;

  // "The onions are chopped"
  until: string;

  // tap callback
  onPress: () => void;
}
export function StepAction(props: StepActionProps) {
  const { timer = false, remaining = null, until, onPress } = props;
  if (remaining !== null && !timer) {
    throw "StepAction: passed time remaining without timer enabled";
  }
  return (
    <Card
      shadowAmt={0.5}
      innerStyle={{
        marginLeft: padding / 2,
        marginRight: padding / 2,
        borderRadius: 10
      }}
    >
      <TouchableOpacity
        onPress={onPress}
        style={{ marginLeft: -padding / 2, marginRight: -padding / 2 }}
      >
        <LeftLine
          hidden
          aside={
            <View
              style={{
                flex: 1,
                display: "flex",
                flexDirection: "column",
                justifyContent: "center"
              }}
            >
              <MaterialIcons
                name={
                  timer
                    ? remaining != null
                      ? "timer"
                      : "play-circle-outline"
                    : "arrow-forward"
                }
                size={40}
              />
            </View>
          }
        >
          <Padded right vertical>
            {remaining != null && (
              <Text style={{ fontSize: 30 }}>
                {stringFormatInterval(remaining)}
              </Text>
            )}
            <View
              style={{
                display: "flex",
                flexDirection: "row",
                alignItems: "center"
              }}
            >
              {remaining != null && (
                <MaterialIcons
                  name="arrow-forward"
                  size={20}
                  style={{ marginRight: 5 }}
                />
              )}
              <Text style={{ fontSize: 20 }}>{until}</Text>
            </View>
          </Padded>
        </LeftLine>
      </TouchableOpacity>
    </Card>
  );
}

export interface Props {
  step: Step;
  num: number;
}

export default function StepView(props: Props) {
  const { step, num } = props;
  return (
    <LeftLine aside={<Circled>{num.toString()}</Circled>}>
      <Padded right bottom>
        <View
          style={{
            minHeight: topSize,
            display: "flex",
            flexDirection: "column",
            justifyContent: "center"
          }}
        >
          <Text style={{ fontSize: 20 }}>{step.name}</Text>
        </View>
        {step.body ? (
          <Markdown>{step.body}</Markdown>
        ) : (
          <View style={{ height: padding }} />
        )}
      </Padded>
    </LeftLine>
  );
}
