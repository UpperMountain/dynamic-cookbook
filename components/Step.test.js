import React from "react";
import renderer from "react-test-renderer";
import StepView from "./Step";

describe("Step", () => {
  it("should render a step with no leading line", () => {
    const cmp = renderer.create(
      <StepView number={1} title={"step one"} body={"this is test body text"} />
    );
    expect(cmp.toJSON()).toMatchSnapshot();
  });

  it("should render a step that does have leading line", () => {
    const cmp = renderer.create(
      <StepView
        number={2}
        title={"step two"}
        body={"this step is not a first step so it has a line leading in"}
      />
    );
    expect(cmp.toJSON()).toMatchSnapshot();
  });
});
