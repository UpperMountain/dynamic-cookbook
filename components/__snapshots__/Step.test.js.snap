// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step should render a step that does have leading line 1`] = `
<View
  style={
    Object {
      "marginTop": 7,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginLeft": 20,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "marginRight": 20,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "borderColor": "black",
              "borderRadius": 15,
              "borderWidth": 1,
              "height": 30,
              "marginBottom": 7,
              "width": 30,
            },
            null,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
                "lineHeight": 28,
                "textAlign": "center",
              },
              Object {
                "color": "black",
              },
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "borderLeftWidth": 1,
              "flex": 1,
              "height": 10,
              "width": 0,
            },
            null,
          ]
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 0.8,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "raleway-medium",
              "fontSize": 24,
            },
            Object {
              "marginBottom": 10,
            },
          ]
        }
      >
        step two
      </Text>
      <Text
        style={
          Object {
            "marginBottom": 10,
          }
        }
      >
        this step is not a first step so it has a line leading in
      </Text>
    </View>
  </View>
  <View>
    <View
      style={
        Array [
          Object {
            "borderLeftWidth": 1,
            "height": 18,
            "width": 0,
          },
          null,
          Object {
            "marginLeft": 34.5,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`Step should render a step with no leading line 1`] = `
<View
  style={
    Object {
      "marginTop": 7,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginLeft": 20,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "marginRight": 20,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "borderColor": "black",
              "borderRadius": 15,
              "borderWidth": 1,
              "height": 30,
              "marginBottom": 7,
              "width": 30,
            },
            null,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
                "lineHeight": 28,
                "textAlign": "center",
              },
              Object {
                "color": "black",
              },
            ]
          }
        >
          1
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "borderLeftWidth": 1,
              "flex": 1,
              "height": 10,
              "width": 0,
            },
            null,
          ]
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 0.8,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "raleway-medium",
              "fontSize": 24,
            },
            Object {
              "marginBottom": 10,
            },
          ]
        }
      >
        step one
      </Text>
      <Text
        style={
          Object {
            "marginBottom": 10,
          }
        }
      >
        this is test body text
      </Text>
    </View>
  </View>
  <View>
    <View
      style={
        Array [
          Object {
            "borderLeftWidth": 1,
            "height": 18,
            "width": 0,
          },
          null,
          Object {
            "marginLeft": 34.5,
          },
        ]
      }
    />
  </View>
</View>
`;
